### ISSUE TITLE
todo list

### ISSUE THREAD

**Issue by 406953373** [406953373/ww_sam_private#10]
【机器学习入门、进阶、深入指南】每一阶段必读论文+课程+代码链接.pdf

--------------------------------------------------------------------
通俗理解条件熵
http://blog.csdn.net/xwd18280820053/article/details/70739368

基于情感词典的文本情感分类
https://www.jianshu.com/p/6826672a379c

偏置-方差分解(Bias-Variance Decomposition)
https://www.cnblogs.com/hithink/p/7372470.html

交叉熵代价函数

GBDT
http://blog.csdn.net/gavin__zhou/article/details/71156715

使用word2vec对新浪微博进行情感分析和分类
http://blog.csdn.net/liugallup/article/details/51164962

《使用sklearn进行集成学习——理论》

http://www.cnblogs.com/jasonfreak/p/5657196.html

《使用sklearn进行集成学习——实践》
http://www.cnblogs.com/jasonfreak/p/5720137.html

GBDT 粗俗讲解
http://blog.csdn.net/w28971023/article/details/8240756

GBDT算法原理深入解析
https://www.zybuluo.com/yxd/note/611571#gbdt%E7%AE%97%E6%B3%95

MLE的解释：王相及的回答
https://www.zhihu.com/question/20447622

PCA Autorencoder

SVM 调参
http://blog.csdn.net/sun_shengyun/article/details/55669160
-------------------------------------
kaggle experience:
feature engineering:
https://www.kaggle.com/jagangupta/stop-the-s-toxic-comments-eda
keras beginners:
https://www.kaggle.com/sbongo/for-beginners-tackling-toxic-using-keras
NLP key papers:
http://www.zhuanzhi.ai/document/0b18475b4fa55a745c9a46adfa3005e2
OOF stacking regime:
https://www.kaggle.com/hhstrand/oof-stacking-regime/code

Linear Baselines:
-Classifying multi-label comments with Logistic Regression
https://www.kaggle.com/rhodiumbeng/classifying-multi-label-comments-0-9741-lb
-NB-SVM strong linear baseline
https://www.kaggle.com/jhoward/nb-svm-strong-linear-baseline

DNN:
Capsule Network:
https://www.kaggle.com/chongjiujjin/capsule-net-with-gru
Keras Text Library(For Reference Only ):
https://github.com/raghakot/keras-text

Language Model:
https://blog.csdn.net/longxinchen_ml/article/details/50646528
NLP+词法系列
https://blog.csdn.net/sinat_26917383/article/details/52275328
NLP+语义分析
https://blog.csdn.net/sinat_26917383/article/details/55683599
-----------------------------
通过kaggle比赛学习机器学习文本分类方法
------------------------------
Classification:
LR
FTLR
XGBOOST
LightGBM
GBDT
FM
FFM
NN

Ensemble:
Stacking

Spark-scala
Hive-hivesql

Github上大神写的相关框架，专门用来处理此类问题：处理样本不均衡的问题
https://github.com/scikit-learn-contrib/imbalanced-learn
under-sampling:

over-sampling:
SMOTE
------------------------
作者：谢志宁
链接：https://www.zhihu.com/question/46688107/answer/117448674
来源：知乎
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

LSTM：像RNN、LSTM、BILSTM这些模型，它们在序列建模上很强大，它们能够capture长远的上下文信息，此外还具备神经网络拟合非线性的能力，这些都是crf无法超越的地方，对于t时刻来说，输出层y_t受到隐层h_t（包含上下文信息）和输入层x_t（当前的输入）的影响，但是y_t和其他时刻的y_t`是相互独立的，感觉像是一种point wise，对当前t时刻来说，我们希望找到一个概率最大的y_t，但其他时刻的y_t`对当前y_t没有影响，如果y_t之间存在较强的依赖关系的话（例如，形容词后面一般接名词，存在一定的约束），LSTM无法对这些约束进行建模，LSTM模型的性能将受到限制。

CRF：它不像LSTM等模型，能够考虑长远的上下文信息，它更多考虑的是整个句子的局部特征的线性加权组合（通过特征模版去扫描整个句子）。关键的一点是，CRF的模型为p(y | x, w)，注意这里y和x都是序列，它有点像list wise，优化的是一个序列y = (y1, y2, …, yn)，而不是某个时刻的y_t，即找到一个概率最高的序列y = (y1, y2, …, yn)使得p(y1, y2, …, yn| x, w)最高，它计算的是一种联合概率，优化的是整个序列（最终目标），而不是将每个时刻的最优拼接起来，在这一点上CRF要优于LSTM。

HMM：CRF不管是在实践还是理论上都要优于HMM，HMM模型的参数主要是“初始的状态分布”，“状态之间的概率转移矩阵”，“状态到观测的概率转移矩阵”，这些信息在CRF中都可以有，例如：在特征模版中考虑h(y1), f(y_i-1, y_i), g(y_i,x_i)等特征。

CRF与LSTM：从数据规模来说，在数据规模较小时，CRF的试验效果要略优于BILSTM，当数据规模较大时，BILSTM的效果应该会超过CRF。从场景来说，如果需要识别的任务不需要太依赖长久的信息，此时RNN等模型只会增加额外的复杂度，此时可以考虑类似科大讯飞FSMN（一种基于窗口考虑上下文信息的“前馈”网络）。

CNN＋BILSTM＋CRF：这是目前学术界比较流行的做法，BILSTM＋CRF是为了结合以上两个模型的优点，CNN主要是处理英文的情况，英文单词是由更细粒度的字母组成，这些字母潜藏着一些特征（例如：前缀后缀特征），通过CNN的卷积操作提取这些特征，在中文中可能并不适用（中文单字无法分解，除非是基于分词后），这里简单举一个例子，例如词性标注场景，单词football与basketball被标为名词的概率较高， 这里后缀ball就是类似这种特征。

BILSTM+CRF的Tensorflow版本：https://github.com/chilynn/sequence-labeling，主要参考了GitHub - glample/tagger: Named Entity Recognition Tool的实现，tagger是基于theano实现的，每一轮的参数更新是基于一个样本的sgd，训练速度比较慢。sequence-labeling是基于tensorflow实现的，将sgd改成mini-batch sgd，由于batch中每个样本的长度不一，训练前需要padding，最后的loss是通过mask进行计算（根据每个样本的真实长度进行计算）。
